### PR TITLE
support dual versions

### PR DESCRIPTION
**Pull request by cheesehary**
Support dual package versions. This can be configured in root package.json

### CODE REVIEW

File: packages/cli/src/checks/__tests__/EXTERNAL_MISMATCH.ts [hunk 1]
```diff
@@ -224,9 +224,38 @@ it("should not error if the value is not a valid semver range", () => {
...
 
-  let errors = internalMismatch.validate(pkg2, ws, rootWorkspace, {});
+  let errors = externalMismatch.validate(pkg2, ws, rootWorkspace, {});
   expect(errors.length).toEqual(0);
 
-  errors = internalMismatch.validate(ws.get("pkg-1")!, ws, rootWorkspace, {});
+  errors = externalMismatch.validate(ws.get("pkg-1")!, ws, rootWorkspace, {});
+  expect(errors.length).toEqual(0);
+});
+
```
Comment (ReDrUm) [inline packages/cli/src/checks/__tests__/EXTERNAL_MISMATCH.ts:233]: We should add a test case to validate which version is used when adding an unsupported version in the case where 2 or more allowed versions exist.
e.g.

```
'allowedDependencyVersions' {
     'foo': [
         '^1.0.0',
         '^2.0.0'
      ]
}
```
`yarn add foo@^3.0.0` ... does it clamp to the lowest version number, the highest one, or the most commonly used between the allowed versions?

Comment (cheesehary) [reply packages/cli/src/checks/__tests__/EXTERNAL_MISMATCH.ts:233]: yep, added a test case, should be the most commonly used one.

File: packages/cli/src/checks/__tests__/EXTERNAL_MISMATCH.ts [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import internalMismatch from "../EXTERNAL_MISMATCH";
```
Comment (ReDrUm) [inline packages/cli/src/checks/__tests__/EXTERNAL_MISMATCH.ts]: Interesting to see the naming discrepancy. Increases the diff, which might be problematic if we did end up sticking to a fork if the contribution to the main lib isn't accepted.

File: packages/cli/src/checks/__tests__/EXTERNAL_MISMATCH.ts [hunk 3]
```diff
@@ -224,9 +224,82 @@ it("should not error if the value is not a valid semver range", () => {
...
+  errors = externalMismatch.validate(
+    ws.get("pkg-1")!,
+    ws,
+    rootWorkspace,
+    options
+  );
+  expect(errors.length).toEqual(0);
+});
+
+it("should error if the range is the range is outside allowedDependencyVersions and running fix should clamp it to the most commonly used one", () => {
```
Comment (ReDrUm) [inline packages/cli/src/checks/__tests__/EXTERNAL_MISMATCH.ts]: grammar nit: "if the range is the range is outside" looks like duplication. Rephrase please.

File: packages/cli/src/checks/utils.ts [hunk 4]
```diff
@@ -120,7 +124,7 @@ function weakMemoize<Arg, Ret>(func: (arg: Arg) => Ret): (arg: Arg) => Ret {
 export let getMostCommonRangeMap = weakMemoize(function getMostCommonRanges(
   allPackages: Map<string, Package>
 ) {
-  let dependencyRangesMapping = new Map<string, {[key: string]: number}>();
+  let dependencyRangesMapping = new Map<string, { [key: string]: number }>();
```
Comment (ReDrUm) [inline packages/cli/src/checks/utils.ts]: We may as well avoid formatting changes that aren't related to the tangible changes of this PR just to reduce diff.

File: packages/cli/src/checks/EXTERNAL_MISMATCH.ts [hunk 5]
```diff
@@ -15,7 +15,7 @@ type ErrorType = {
 };
 
 export default makeCheck<ErrorType>({
-  validate: (workspace, allWorkspace) => {
+  validate: (workspace, allWorkspace, rootWorkspace, options) => {
```
Comment (marcodejongh) [inline packages/cli/src/checks/EXTERNAL_MISMATCH.ts:18]: What does rootWorkspace do here? I don't see it referenced further down?

File: packages/cli/src/checks/EXTERNAL_MISMATCH.ts [hunk 6]
```diff
@@ -25,9 +25,13 @@ export default makeCheck<ErrorType>({
         for (let depName in deps) {
           let range = deps[depName];
           let mostCommonRange = mostCommonRangeMap.get(depName);
+          const allowedVersions =
+            options.allowedDependencyVersions &&
+            options.allowedDependencyVersions[depName];
           if (
             mostCommonRange !== undefined &&
             mostCommonRange !== range &&
+            !(allowedVersions && allowedVersions.includes(range)) &&
```
Comment (marcodejongh) [inline packages/cli/src/checks/EXTERNAL_MISMATCH.ts:34]: Just to summarise, it'll error anything that isn't most common or in allowedVersions? When fixing it'll fix to most common?

File: packages/cli/src/checks/EXTERNAL_MISMATCH.ts (hunk 5 above)
Comment (cheesehary) [reply packages/cli/src/checks/EXTERNAL_MISMATCH.ts:18]: That's the validate function signature.
`let errors = check.validate(
          workspace,
          allWorkspaces,
          rootWorkspace,
          options
        );`

File: packages/cli/src/checks/EXTERNAL_MISMATCH.ts (hunk 6 above)
Comment (cheesehary) [reply packages/cli/src/checks/EXTERNAL_MISMATCH.ts:34]: yep correct

File: packages/cli/src/checks/EXTERNAL_MISMATCH.ts (hunk 5 above)
Comment (marcodejongh) [reply packages/cli/src/checks/EXTERNAL_MISMATCH.ts:18]: Ahhh right

**Review by marcodejongh — APPROVED** [review]

File: packages/cli/src/checks/utils.ts (hunk 4 above)
Comment (cheesehary) [reply packages/cli/src/checks/utils.ts]: yeah, I cleared all those unrelated changes to reduce diff

**Comment by marcodejongh** [comment]
Dont forget to open a PR back upstream

**Comment by ReDrUm** [comment]
Late to the party, but it seems like we may need to evolve this a little further first. There seems to be unexpected changes to the `yarn.lock` file. See this comment thread https://bitbucket.org/atlassian/atlassian-frontend/pull-requests/18569#comment-277298005

**Pull request merged**